spices/SPICE-0001-import-graph-analyzer-api.adoc

```diff
@@ -0,0 +1,336 @@
+  ["file:///foo/bar/Bar.pkl"] = "file:///foo/bar/bar.pkl" // <3>
+  ["package://example.com/bar@1.0.0#/Bar.pkl"] = "package://example.com/bar@1.0.0#/Bar.pkl" // <4>
+  ["pkl:reflect"] = "pkl:reflect"
+  ["pkl:jsonnet"] = "pkl:jsonnet"
+  ["pkl:xml"] = "pkl:xml"
+  ["pkl:protobuf"] = "pkl:protobuf"
+  ["file:///path/to/foo.pkl"] = "file:///path/to/foo.pkl"
+  ["file:///path/to/my/foo1.pkl"] = "file:///path/to/my/foo1.pkl"
+  ["file:///path/to/my/foo2.pkl"] = "file:///path/to/my/foo2.pkl"
+  ["file:///path/to/my/subdir/foo.pkl"] = "file:///path/to/my/subdir/foo.pkl"
+}
+----
+<1> dependency notation turns into `projectpackage://`.
+<2> local dependency's resolve URI is a local file.
+<3> module URIs that don't resolve to anything further stay the same.
+<4> package URIs stay the same.
+
+=== Gradle Plugin Changes
+
+The input files for various Pkl Gradle tasks will switch to using the Java `ImportAnalyzer` to build its input files.
+
+The set of input files will be `resolvedUris` whose keys start with `\file:///`.
+All other modules are assumed to be external to the machine and cache-able.
+
+The `transitiveModules` property is removed.
+
+== Future directions
+
+=== `pkl:anaylze` standard library module
+
+This is a starting point for the `pkl.analyze` stdlib module.
+Other forms of static analysis can be added to it as the need grows.
+
+=== Support in other language bindings
+
+This feature introduces `PklAnalyzer`, a new Java API that can analyze the imports of modules without requiring the evaluator.
+This is a capability that isn't being added to other language bindings.
+
+A simple workaround is to use the standard evaluator API, given the introduction of the new `pkl.anaylze` stdlib module, albeit with the drawback of the overhead of performing Pkl evaluation.
+
+To support evaluator-free analysis, a future improvement can be to add messages to the server API for performing import graph analysis.
+
+== Alternatives considered
+
+=== Enhance `pkl.reflect`
+
+One alternative is to enhance the `pkl.reflect` API to provide the import graph.
+
+Imagined API:
+
+[source,diff]
+----
+ module pkl.reflect
+
+ class Module {
++  /// The imports, extends, and amends declared by this module.
++  imports: Listing<Module>
+ }
+----
+
+The benefit of this approach is that an import can be defined in terms of regular Pkl evaluation.
+We do not need to provide a new CLI command, nor extension points in Gradle or server mode.
+
+Some drawbacks:
+
+* Reflection currently only provides information about a module's schema, using metadata that is already gathered when a module is initialized. Exposing this level of import requires that _data_ is also gathered during module initialization. This adds overhead to basic evaluation, even if the import graph is not needed.
+* Using `pkl:reflect` requires that Pkl initializes a module, which is much more expensive than just parsing it. Some modules are known to take seconds to initialize.
```

Comment:
What exactly is module initialization, and why can it take seconds?

Reply:
Module initialization is the first thing that happens when you do evaluation. The module's class is initialized, and then Pkl will execute something within the module (e.g. read `output.text`).

When initializing a class, Pkl will load all class properties, which is eager. When Pkl loads this following module, it also executes imports because they are required to initialize the types:

```
import "Bar.pkl"

bar: Bar
```

Overall, this can be expensive in large enough modules.

Reply:
Does initialization include parsing? And what does loading a property entail?